slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -92,6 +93,47 @@ ::grpc::Status SLAMServiceImpl::GetPositionNew(
     return grpc::Status::OK;
 }
 
+::grpc::Status SLAMServiceImpl::GetPointCloudMap(
+    ServerContext *context, const GetPointCloudMapRequest *request,
+    GetPointCloudMapResponse *response) {
+    std::string pointcloud_map;
+    // Write or grab the latest pointcloud map in form of a string
+    try {
+        std::shared_lock optimization_lock{optimization_shared_mutex,
+                                           std::defer_lock};
+        if (optimization_lock.try_lock()) {
+            // We are able to lock the optimization_shared_mutex, which means
+            // that the optimization is not ongoing and we can grab the newest
+            // map
+            GetLatestSampledPointCloudMapString(pointcloud_map);
```

Comment:
Is there any reason why we don't set `latest_pointcloud_map` in this case?

If I'm understanding this code correctly, the way it is currently written its possible a caller to get cached data that is older than the last GetPointCloudMap call they performed.

That can happen if:
1. the cached data is from time T1
2. `GetPointCloudMap` is called & the `optimization_lock` is acquired. A map from time T2 is returned (but not cached)
3. `GetPointCloudMap`  is called & the `optimization_lock` is NOT acquired (as it is held by the data thread). The cached data is returned from T1 which is OLDER than the data the client just received. 

If I were a client, I would not be satisfied with sometimes getting a map older than the map I was just given. 

I would expect `GetPointCloudMap` to either:
1. Give me the map it just gave me
2. Give me a newer, better map


@tessavitabile please correct me if I'm mistaken about any of the above.

Reply:
Also, as a note for the future (no change needed here): when writing functions that only exist to return a string (like this one does), we should stop writing functions which take a string, mutate the string & return void. Instead, we should just return the string by value.  

From reading the Tour of C++ book's section on strings, the creator of C++ recommends (section 10.2) returning strings by value, rather than mutating a string argument & relying on the caller to know that the argument now contains the data supplied by the function. Also, the efficiency of both approaches should be the same as frequently no copy is performed when returning a string by value, as std::sting implement the move constructor (and also b/c apparently, the compiler is able to use copy elusion to even avoid the move). 

@tessavitabile Please correct me if any of that is inaccurate (trying to get my C++ legs here).

Reply:
I think you're right that there is a race condition where optimization caches a map, then GetPointCloudMap creates and returns a new map, then GetPointCloudMap returns the cached map. However, no changes would occur to the map during this time because optimization only runs at the end of processing in offline mode (and also at the beginning when loading an a priori map, but GetPointCloudMap will not return anything at this time). So I don't think we'll actually see maps go backward in time.

Yes, it is better to return a string by value.

Reply:
Tess beat me to it, but yeah this optimization lock should only be NOT acquired in offline mode at the end of a run or on startup if a flag is enabled. 

As far as setting `latest_pointcloud_map` is concerned I personally don't see the harm in it, so I'll set it up in the relevant functions

Reply:
cool, thanks for explaining! I missed the optimization step nuance.